README.md
# Platform Science Code Exercise 1

---

[![Run Tests](https://github.com/patinthehat/ps-code-exercise-01/actions/workflows/run-tests.yml/badge.svg)](https://github.com/patinthehat/ps-code-exercise-01/actions/workflows/run-tests.yml) [![codecov](https://codecov.io/gh/patinthehat/ps-code-exercise-01/branch/main/graph/badge.svg?token=BZenrQO00O)](https://codecov.io/gh/patinthehat/ps-code-exercise-01)

## Overview

This exercise takes input of a list of addresses and drivers, and outputs the best driver for each address. Once a driver has been assigned to a shipment, they cannot be assigned to another shipment. Assignments are done sequentially based on shipment (address).

Additionally, it also can:

-   List the top five drivers for a given address
-   List the best driver for each shipment without considering previous shipments

It is a command line application, written in TypeScript and compiled with [esbuild](https://esbuild.github.io/). See below for instructions on how to run the application.

## Assumptions

This code exercise assumes the following:

-   The address/shipment files contain a single address per line, in this format: `street, city, state postalCode`
-   The address/shipment items contain state abbreviations only, not full state names (e.g. `CA` instead of `California`)
-   The address/shipment items are US addresses only and postal codes are 5 digits
-   The driver file contains a single driver per line, in this format: `firstName lastName`

## Setup

```bash
npm install
```

## Running the application

To run the default code exercise, run the following command:

```bash
npm run dev
```

This will run the default code exercise with test datasets.

Additionally, there are three commands available:

-   `route-shipments`
-   `best-driver`
-   `top-five`

```bash
# run one of the available commands with the following command:
npm run dev -- [command_name] [addresses_filename] [drivers_filename]
```

The previous command will compile the TypeScript code with `esbuild` and run the application using the provided parameters. Filenames should be absolute file paths.

## Testing

This project uses Jest for unit tests. To run the test suite, run:

```bash
npm run test
```

To display the code coverage for the entire application at the file level, run:

```bash
npm run test:coverage
```

### Coverage

Sunburst:

![Coverage Sunburst](https://codecov.io/gh/patinthehat/ps-code-exercise-01/branch/main/graphs/sunburst.svg?token=BZenrQO00O)

Icicle:

![Coverage Icicle](https://codecov.io/gh/patinthehat/ps-code-exercise-01/branch/main/graphs/icicle.svg?token=BZenrQO00O)

---

## Credits

-   [Patrick Organ](https://github.com/patinthehat)

## License

The MIT License (MIT). Please see [License File](LICENSE) for more information.


CHANGELOG.md
# Changelog

All notable changes to `ps-code-exercise-01` will be documented in this file.

---

## 1.0.0 - unreleased

-   initial release


tests/lib/helpers.test.ts
import { calculateStringStatistics, normalizeStreetAddress, readTextFileLines } from '@/lib/helpers';

it('calculates the correct number of vowels in a string', () => {
    expect(calculateStringStatistics('hello').vowels).toBe(2);
});

it('calculates the correct number of consonants in a string', () => {
    expect(calculateStringStatistics('hello').consonants).toBe(3);
});

it('calculates the correct number of vowels + consonants in a string', () => {
    const stats = calculateStringStatistics('hello');
    expect(stats.consonants + stats.vowels).toBe(5);
});

it('calculates the correct length of a string', () => {
    expect(calculateStringStatistics('hello').length).toBe(5);
});

it('reads the correct number of lines from a text file', () => {
    expect(readTextFileLines(`${__dirname}/../fixtures/drivers-01.txt`).length).toBe(15);
});

it('reads the correct contents from a text file', () => {
    expect(readTextFileLines(`${__dirname}/../fixtures/drivers-01.txt`)).toMatchSnapshot();
});

it('normalizes street addresses', () => {
    const addresses = ['123 Main St.', '123 Main Ave.', '1231 Yonge Street', '456 West 1st Rd.', '9371 Eastwood Avenue'];

    const expectedAddress = ['123 Main Street', '123 Main Avenue', '1231 Yonge Street', '456 West 1st Road', '9371 Eastwood Avenue'];

    addresses.forEach((address, index) => {
        expect(normalizeStreetAddress(address)).toBe(expectedAddress[index]);
    });
});


tests/lib/Shipment.test.ts
/* eslint-disable array-element-newline */
import { Driver } from '@/lib/Driver';
import { Shipment } from '@/lib/Shipment';

it('calculates the correct suitability score for a shipment and driver', () => {
    const shipment = new Shipment('9 Olive Avenue');
    const driver = new Driver('Malcolm Parker');

    expect(shipment.suitabilityScoreForDriver(driver)).toEqual(20.25);
});

it('finds the best driver for a shipment', () => {
    const shipment = new Shipment('9 Olive Avenue');

    const drivers = [new Driver('Zachariah Hardwick'), new Driver('Malcolm Parker'), new Driver('Lillian Mccormick')];

    expect(shipment.getBestDriver(drivers).name).toEqual(drivers[1].name);
});

it('throws an error when no drivers are provided', () => {
    const shipment = new Shipment('9 Olive Avenue');

    expect(() => {
        shipment.getBestDriver([]);
    }).toThrowErrorMatchingSnapshot();
});

it('gets the state from the address', () => {
    const shipment = new Shipment('9 Olive Avenue, Somewhere, CA 90210');

    expect(shipment.state).toEqual('CA');
});

it('gets the postal code from the address', () => {
    const shipment = new Shipment('9 Olive Avenue, Somewhere, CA 90210');

    expect(shipment.postalCode).toEqual('90210');
});

it('gets the city from the address', () => {
    const shipment = new Shipment('9 Olive Avenue, Somewhere, CA 90210');

    expect(shipment.city).toEqual('Somewhere');
});


tests/lib/ShipmentManager.test.ts
/* eslint-disable array-element-newline */
import { Driver } from '@/lib/Driver';
import { Shipment } from '@/lib/Shipment';
import { ShipmentManager } from '@/lib/ShipmentManager';

it('returns the correct driver and score for a shipment', () => {
    const shipments = [new Shipment('9 Olive Avenue'), new Shipment('7192 Glendale Lane'), new Shipment('1 Main Street')];

    const drivers = [new Driver('Malcolm Parker'), new Driver('Zachariah Hardwick'), new Driver('Lillian Mccormick')];

    const mgr = new ShipmentManager(shipments, drivers);

    mgr.assignShipmentsToDrivers();

    expect(mgr.assignments[0].shipment.address).toEqual('9 Olive Avenue');
    expect(mgr.assignments[0].driver.name).toEqual('Malcolm Parker');
    expect(mgr.assignments[0].suitabilityScore).toEqual(20.25);

    expect(mgr.assignments[1].shipment.address).toEqual('7192 Glendale Lane');
    expect(mgr.assignments[1].driver.name).toEqual('Zachariah Hardwick');
    expect(mgr.assignments[1].suitabilityScore).toEqual(68.34375);
});


jest.config.js
const { pathsToModuleNameMapper } = require('ts-jest');
//const { compilerOptions } = require('./tsconfig.json');
const tsConfigPaths = {
    '@/*': ['src/*'],
    '@tests/*': ['tests/*'],
};

/** @type {import('@jest/types').Config.InitialOptions } */
module.exports = {
    preset: 'ts-jest/presets/js-with-ts',
    testEnvironment: 'node',
    transform: { '^.+\\.tsx?$': 'ts-jest' },
    testRegex: '(/__test__/.*|/tests/.*|(\\.|/)(test|spec))\\.[tj]sx?$',
    testPathIgnorePatterns: ['/node_modules/', '/dist/'],
    moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json'],
    moduleNameMapper: pathsToModuleNameMapper(tsConfigPaths, { prefix: `${__dirname}/` }),

    coverageDirectory: './coverage',
    coverageReporters: ['html', 'text'],
    collectCoverageFrom: ['src/**/*.{ts,js}', '!**/node_modules/**', '!**/vendor/**', '!**/dist/**', '!**/tests/**', '!src/index.ts'],
};


src/lib/helpers.ts
import { StringStatistics } from '@/types/StringStatistics';
import { readFileSync } from 'fs';

export const readTextFile = (path: string): string => readFileSync(path, { encoding: 'utf8' });

export const readTextFileLines = (path: string): string[] =>
    readTextFile(path)
        .trim()
        .split(/\n/)
        .filter(line => line.trim().length > 0);

export const calculateStringStatistics = (str: string): StringStatistics => {
    const alphabet = 'abcdefghijklmnopqrstuvwxyz';
    const vowels = 'aeiou'.split('');
    const consonants = alphabet.split('').filter(letter => !vowels.includes(letter));

    const result = {
        consonants: str.split('').filter(letter => consonants.includes(letter)).length,
        length: str.length,
        vowels: str.split('').filter(letter => vowels.includes(letter)).length,
    };

    return result;
};

export const normalizeStreetAddress = (address: string): string => {
    return address
        .replace(/ Ave\./i, ' Avenue')
        .replace(/ Blvd\./i, ' Boulevard')
        .replace(/ Cir\./i, ' Circle')
        .replace(/ Ct\./i, ' Court')
        .replace(/ Dr\./i, ' Drive')
        .replace(/ Hwy\./i, ' Highway')
        .replace(/ Ln\./i, ' Lane')
        .replace(/ Pkwy\./i, ' Parkway')
        .replace(/ Pl\./i, ' Place')
        .replace(/ Rd\./i, ' Road')
        .replace(/ St\./i, ' Street');
};


src/lib/Shipment.ts
import { Driver } from '@/lib/Driver';
import { calculateStringStatistics, normalizeStreetAddress } from '@/lib/helpers';

export class Shipment {
    public streetAddress = '';

    constructor(public address: string) {
        this.streetAddress = this.getStreetAddress();
    }

    public getStreetAddress(): string {
        const result = this.address.split(',')[0].trim();
        return normalizeStreetAddress(result);
    }

    public get city(): string {
        return this.address.split(',')[1].trim();
    }

    public get state(): string {
        const match = this.address.match(/([A-Z]{2})\s+\d{5}$/);
        if (match) {
            return match[1];
        }

        return 'UNKNOWN';
    }

    public get postalCode(): string {
        const match = this.address.match(/[A-Z]{2}\s+(\d{5})$/);
        if (match) {
            return match[1];
        }

        return '00000';
    }

    public suitabilityScoreForDriver(driver: Driver): number {
        const addressStats = calculateStringStatistics(this.streetAddress);
        const nameStats = calculateStringStatistics(driver.name);

        let baseSuitabilityScore = 0;
        let result = 0;

        // odd
        if (addressStats.length % 2 !== 0) {
            baseSuitabilityScore = nameStats.consonants * 1.0;
        } else {
            baseSuitabilityScore = nameStats.vowels * 1.5;
        }

        result = baseSuitabilityScore;

        // common factors increase score by 50%
        for (let i = 2; i <= addressStats.length; i++) {
            if (addressStats.length % i === 0 && nameStats.length % i === 0) {
                result += result * 0.5;
            }
        }

        return result;
    }

    public getBestDriver(drivers: Driver[]): Driver {
        let bestDriver: Driver;
        let bestScore = 0;

        if (!drivers.length) {
            throw new Error('Cannot calculate best driver: no drivers provided.');
        }

        bestDriver = drivers[0];

        drivers.forEach(driver => {
            const score = driver.suitabilityScoreForShipment(this);

            if (score > bestScore) {
                bestScore = score;
                bestDriver = driver;
            }
        });

        return bestDriver;
    }
}


src/lib/ShipmentManager.ts
import { Driver } from '@/lib/Driver';
import { Shipment } from '@/lib/Shipment';
import { ShipmentAssignment } from '@/types/ShipmentAssignment';

export class ShipmentManager {
    public assignments: ShipmentAssignment[] = [];

    constructor(public shipments: Shipment[], public drivers: Driver[]) {
        //
    }

    public assignShipmentsToDrivers(): void {
        this.assignments = [];

        const scores = {};

        this.drivers.forEach(driver => {
            scores[driver.name] = {};
        });

        this.shipments.forEach(shipment => {
            this.drivers.forEach(driver => {
                scores[driver.name][shipment.address] = driver.suitabilityScoreForShipment(shipment);
            });
        });

        for (const driver of this.drivers) {
            let bestScore = 0;
            let bestShipment: any = null;

            for (const shipment of this.shipments) {
                if (scores[driver.name][shipment.address] > bestScore) {
                    bestScore = scores[driver.name][shipment.address];
                    bestShipment = shipment;
                }
            }

            if (bestShipment) {
                this.assignments.push({
                    driver,
                    shipment: bestShipment,
                    suitabilityScore: bestScore,
                });

                this.shipments = this.shipments.filter(shipment => shipment.address.toLowerCase() !== bestShipment.address.toLowerCase());
            }
        }
    }
}


src/index.ts
import { Driver } from '@/lib/Driver';
import { readTextFileLines } from '@/lib/helpers';
import { Shipment } from '@/lib/Shipment';
import { ShipmentManager } from '@/lib/ShipmentManager';
import { Command } from 'commander';
import { existsSync } from 'fs';
const program = new Command();

/*
The top-secret algorithm is:
● If the length of the shipment's destination street name is even, the base suitability score (SS) is the number of vowels in the driver’s name multiplied by 1.5.
● If the length of the shipment's destination street name is odd, the base SS is the number of consonants in the driver’s name multiplied by 1.

● If the length of the shipment's destination street name shares any common factors (besides 1) with the length of the driver’s name, the SS is increased by 50% above the base SS.

Write an application in the language of your choice that assigns shipment destinations to drivers in a way that maximizes the total SS over the set of drivers.
Each driver can only have one shipment and each shipment can only be offered to one driver.
Your program should run on the command line and take as input two newline separated files, the first containing the street addresses of the shipment destinations and the second containing the names of the drivers.
The output should be the total SS and a matching between shipment destinations and drivers. You do not need to worry about malformed input, but you should certainly handle both upper and lower case names.
*/

program.name('code-exercise-1').version('1.0.0');

program
    .command('route-shipments', { isDefault: true })
    .description('Route shipments for the given input files')
    .argument('[addresses]', 'addresses filename', 'datasets/01/shipments.txt')
    .argument('[driversFile]', 'drivers filename', 'datasets/01/drivers.txt')
    .action((addressesFile, driversFile, options) => {
        if (!existsSync(addressesFile)) {
            addressesFile = `${__dirname}/${addressesFile}`;
        }

        if (!existsSync(driversFile)) {
            driversFile = `${__dirname}/${driversFile}`;
        }

        console.log('----');
        console.log(`Routing shipments for ${addressesFile.replace(__dirname, '.')} and ${driversFile.replace(__dirname, '.')}`);

        const addressesData = readTextFileLines(addressesFile);
        const shipments: Shipment[] = addressesData.map(address => new Shipment(address));

        const driversData = readTextFileLines(driversFile);
        const drivers: Driver[] = driversData.map(driver => new Driver(driver));

        const shipmentManager = new ShipmentManager(shipments, drivers);

        shipmentManager.assignShipmentsToDrivers();

        console.log('----');
        console.log('Driver assignments for each shipment:');
        console.log(
            shipmentManager.assignments
                .map(assignment => `${assignment.shipment.streetAddress} => ${assignment.driver.name} (${assignment.suitabilityScore})`)
                .join('\n'),
        );
    });

program
    .command('best-driver')
    .description('Show the best driver for each shipment')
    .argument('[addresses]', 'addresses filename', 'datasets/01/shipments.txt')
    .argument('[driversFile]', 'drivers filename', 'datasets/01/drivers.txt')
    .action((addressesFile, driversFile) => {
        if (!existsSync(addressesFile)) {
            addressesFile = `${__dirname}/${addressesFile}`;
        }

        if (!existsSync(driversFile)) {
            driversFile = `${__dirname}/${driversFile}`;
        }

        const addressesData = readTextFileLines(addressesFile);
        const shipments: Shipment[] = addressesData.map(address => new Shipment(address));
        const immutableShipments = shipments.slice(0);

        const driversData = readTextFileLines(driversFile);
        const drivers: Driver[] = driversData.map(driver => new Driver(driver));

        console.log('----');
        console.log('Top drivers for each shipment');
        console.log('----');

        immutableShipments.forEach(shipment => {
            const driver = shipment.getBestDriver(drivers);
            console.log(`Shipment: ${shipment.streetAddress}; best driver: ${driver.name} (${shipment.suitabilityScoreForDriver(driver)})`);
        });
    });

program
    .command('top-five')
    .description('Show top five drivers for each shipment')
    .argument('[addresses]', 'addresses filename', 'datasets/01/shipments.txt')
    .argument('[driversFile]', 'drivers filename', 'datasets/01/drivers.txt')
    .action((addressesFile, driversFile) => {
        if (!existsSync(addressesFile)) {
            addressesFile = `${__dirname}/${addressesFile}`;
        }

        if (!existsSync(driversFile)) {
            driversFile = `${__dirname}/${driversFile}`;
        }

        const addressesData = readTextFileLines(addressesFile);
        const shipments: Shipment[] = addressesData.map(address => new Shipment(address));
        const immutableShipments = shipments.slice(0);

        const driversData = readTextFileLines(driversFile);
        const drivers: Driver[] = driversData.map(driver => new Driver(driver));

        console.log('----');
        console.log('Top 5 Shipment/Driver Rankings');
        console.log('----');

        immutableShipments.forEach(shipment => {
            console.log(`Shipment: ${shipment.streetAddress};`);
            const rankedDrivers: any[] = [];

            drivers.forEach(driver => {
                const score = shipment.suitabilityScoreForDriver(driver);
                rankedDrivers.push({ driver, score });
            });

            rankedDrivers.sort((a, b) => b.score - a.score);

            rankedDrivers.slice(0, 5).forEach((ranked: { driver: Driver; score: number }) => {
                console.log(`   - ${ranked.driver.name} (${ranked.score})`);
            });
        });
    });

program.parse();
